Fix #6 by limiting the length of auth principals to the max principal length in a DatalogGuard query.

tmroeder opened the pull request:
These commits add functions to compute the max length of principals in a DatalogGuard query. The DatalogGuard uses these in its subprin/3 implementation to skip (Var, Const, Const) discovery if the principal formed from the two constants has a length greater than the maximum length allowed.

kevinawalsh commented on go/tao/datalog_guard.go:
```diff
@@ -618,7 +624,122 @@ func (g *DatalogGuard) retract(f auth.Form) error {
 }
 
+func max(args ...int) int {
+	m := 0
```
This function is overkill. It should just be a standard 2-arg max function. See below.

kevinawalsh commented on go/tao/datalog_guard.go:
```diff
@@ -618,7 +624,122 @@ func (g *DatalogGuard) retract(f auth.Form) error {
+	switch f.(type) {
+	case auth.Pred:
+		for _, t := range f.(auth.Pred).Arg {
+			m = append(m, getMaxTermLength(t))
```
Make m an int, so here you can do m = max(m, getMaxTermLength(t)). There is no need to keep an array of terms just you can take the max later. Same comment several times below.

kevinawalsh commented on go/tao/datalog_guard.go:
```diff
@@ -618,7 +624,122 @@ func (g *DatalogGuard) retract(f auth.Form) error {
+
+	return max(m...)
+}
+
```
This is all kind of painful for me to read, since I feel responsible for how many cases are here. Should we perhaps include a generic breadth/depth-first traversal function as part of the auth package?

tmroeder commented:
Thanks for the comments. I've fixed the issues with max and added a test that explicitly tries to trigger the infinite loop case in the DatalogGuard.

With respect to added generic traversal in the auth package, I think that should wait until we need to do something like this a second time.

tmroeder commented:
The only outstanding question here is about whether or not we should add a generic tree-walking function in the auth library. I'd like to push that off to another change someday when we need it. Are you OK with this?

kevinawalsh commented:
Agreed.

On Thu, Aug 13, 2015 at 7:02 PM, Tom Roeder <notifications@github.com>
wrote:

> The only outstanding question here is about whether or not we should add a
> generic tree-walking function in the auth library. I'd like to push that
> off to another change someday when we need it. Are you OK with this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jlmucb/cloudproxy/pull/21#issuecomment-130872670>.
>

tmroeder commented:
Fixed #6 with push of rebased version of this branch
